src/org/sosy_lab/common/io/TempFile.java

```diff
@@ -205,4 +196,73 @@ public void close() throws IOException {
       super.close();
     }
   }
+
+  /**
+   * Create a fresh temporary directory in JVM's temp directory.
+   *
+   * <p>The resulting {@link Path} object is wrapped in a {@link DeleteOnCloseDir}, which deletes
+   * the directory recursively including its contents as soon as {@link DeleteOnCloseDir#close()} is
+   * called.
+   *
+   * <p>It is recommended to use the following pattern: <code>
+   * try (DeleteOnCloseDir tempDir = TempFile.createDeleteOnCloseDir()) {
+   *   // use tempDir.toPath() to get the Path object denoting the temporary directory
+   * }
+   * </code>
+   *
+   * @param pPrefix The prefix of the randomly-generated directory name.
+   * @param pFileAttributes The {@link FileAttribute}s used for creating the directory.
+   */
+  public static DeleteOnCloseDir createDeleteOnCloseDir(
+      String pPrefix, FileAttribute<?>... pFileAttributes) throws IOException {
+    checkNotNull(pPrefix);
+    Path tempDir;
+    try {
+      tempDir = Files.createTempDirectory(TMPDIR, pPrefix, pFileAttributes.clone());
+    } catch (IOException e) {
+      throw improveIoExceptionMessage(e, TMPDIR, pPrefix + "*");
+    }
+    return new DeleteOnCloseDir(tempDir);
+  }
+
+  /**
+   * A simple wrapper around {@link Path} that calls {@link
+   * com.google.common.io.MoreFiles#deleteRecursively(Path,
+   * com.google.common.io.RecursiveDeleteOption...)} recursively from {@link AutoCloseable#close()}
+   * to delete the directory including its contents.
+   */
+  @Immutable
+  public static final class DeleteOnCloseDir implements AutoCloseable {
+
+    private final Path path;
+
+    private DeleteOnCloseDir(Path pDir) {
+      path = pDir;
+    }
+
+    public Path toPath() {
+      return path;
+    }
+
+    @Override
+    public void close() throws IOException {
```

Comment:
Given that this deletes data, it might be good to have an explicit JavaDoc with "Recursively delete all files and directories in the directory represented by this instance."